pkg/storage/ingest/reader.go

```diff
@@ -243,18 +302,27 @@ func (r *PartitionReader) processNextFetchesUntilMaxLagHonored(ctx context.Conte
 		}
 
 		if boff.Err() != nil {
-			return boff.Err()
+			// TODO should be moved to dskit's backoff
+			if ctx.Err() != nil {
+				return 0, context.Cause(ctx)
+			}
+
+			return 0, boff.Err()
 		}
 
 		// If it took less than the max desired lag to replay the partition
 		// then we can stop here, otherwise we'll have to redo it.
-		if currLag := time.Since(lastProducedOffsetFetchedAt); currLag <= maxLag {
-			level.Info(r.logger).Log("msg", "partition reader consumed partition and current lag is less than configured max consumer lag", "last_consumed_offset", r.consumedOffsetWatcher.LastConsumedOffset(), "current_lag", currLag, "max_lag", maxLag)
-			return nil
+		if currLag = time.Since(lastProducedOffsetRequestedAt); currLag <= maxLag {
+			return currLag, nil
 		}
 	}
 
-	return boff.Err()
+	// TODO should be moved to dskit's backoff
```

Comment:
Note to reviewers: I will open a PR to add `context.Cause()` support to dskit's `backoff` but I prefer to not block on it.

Reply:
See: https://github.com/grafana/dskit/pull/538